[lldb]Implement LLDB Telemetry

oontvoo opened the pull request:
Provide some concrete implementation for llvm/Telemetry in lldb/Telemetry

(Dependent on [PR/102323](https://github.com/llvm/llvm-project/pull/102323)

nikic commented:
Why does any part of this need to be in llvm/ at all?

oontvoo commented:
> Why does any part of this need to be in llvm/ at all?

Please see comments on the rfc at https://discourse.llvm.org/t/rfc-lldb-telemetry-metrics/64588/15

TL;DR : A few people feel that Telemetry is not something unique to LLDB and that other tools might be interested in having it too. So rather than having mutliple Telemetry libs doing similar things, it might be good to have at least a common API

labath reviewed (commented):
I've made one pass over the PR. I think this would be easier to review if it was split into multiple patches:
- core llvm infrastructure
- core lldb infrastructure
- ~one patch per event type

I think the biggest hurdle will be finding someone to approve the addition of the new llvm library (I doubt it's going to be someone from the current reviewer set). I wouldn't be surprised if this ends up needing its own RFC.

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,153 @@
+struct TargetTelemetryInfo : public ::llvm::telemetry::TelemetryInfo {
+  // All entries emitted for the same SBTarget will have the same
+  // target_uuid.
+  std::string target_uuid;
```
This UUID is the same as the executable module's UUID.
If the process does execve, then a new set of TargetTelemetryInfo will be created for the new UUID.

labath replied:
That's probably fine, but then this comment is not correct, as this will still be the same (SB)Target object (only it's executable module will have changed)

oontvoo commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,619 @@
+
+  auto *ret = new llvm::telemetry::TelemetryConfig{enable_telemetry,
+                                                   additional_destinations};
+#ifdef HAS_VENDOR_TELEMETRY_CONFIG
```
can you clarify what you mean by dummy implementations? (w/o causing either duplicate-symbols errors) I'd tried the weak-def but it wasn't very portable, so I thought this was the simplest.

labath replied:
I just meant that we could conditionally provide an empty implementation of the "vendor" function instead of conditionally calling it. So something like:
```
#if VENDOR
// do something to get the vendor definition
#else
// provide an empty definition ourselves, e.g:
void ApplyVendorSpecificConfigs(...) {}
#endif
```

Then you don't need the ifdef at the call site.

oontvoo replied:
Done - reworked this to use Plugin registry (the default plugin is enabled for test, then vendors can add their own impl as needed)

oontvoo commented:
> I've made one pass over the PR. I think this would be easier to review if it was split into multiple patches:
> 
> * core llvm infrastructure
> * core lldb infrastructure
> * ~one patch per event type
> 
> I think the biggest hurdle will be finding someone to approve the addition of the new llvm library (I doubt it's going to be someone from the current reviewer set). I wouldn't be surprised if this ends up needing its own RFC.

Good point - so here's the plane:
 - I'll create a separate PR for the first bullet point (llvm core infra).
 - keep this PR as is - since branching is such a pain (and controversial) in llvm, you can "pretend" to not see the llvm/lib/ changes in this PR . hopefully the diff will disappear once the first PR goes in.

oontvoo committed:
[llvm][lib]Propose a simple Telemetry framework.

Objective:
  - Provide a common framework in LLVM for collecting various usage metrics
  - Characteristics:
      + Extensible and configurable by:
          * tools in LLVM that want to use it
          * vendors in their downstream codebase
          * tools users (as allowed by vendor)

Background:
The framework was originally proposed only for LLDB, but there were quite a few requests
that it should be moved to llvm/lib given telemetry is a common usage to a lot of tools,
not just LLDB.

See more details on the design and discussions here on the RFC: https://discourse.llvm.org/t/rfc-lldb-telemetry-metrics/64588/20?u=oontvoo

oontvoo committed:
add header

oontvoo committed:
fixed typo

oontvoo committed:
Merge branch 'llvm:main' into llvm_telemetry

oontvoo committed:
Merge branch 'llvm:main' into llvm_telemetry

oontvoo committed:
add tests and addressed review comments

oontvoo committed:
formatting changes

oontvoo committed:
more formatting

oontvoo committed:
more formatting changes

oontvoo committed:
Added header comment to describe the package

oontvoo committed:
reformat header

oontvoo committed:
addressed review comments and added separate docs in llvm/docs/

oontvoo committed:
updated field names in tests

oontvoo committed:
reformated doc and added additional details

oontvoo committed:
fix formatting

oontvoo committed:
details

oontvoo committed:
formatted TelemetryTest.cpp

oontvoo committed:
fix formatting in docs

oontvoo committed:
convert comments to doxygen style

oontvoo committed:
fix doc toc

oontvoo committed:
group all the testing-params into a Context struct

oontvoo committed:
fix conversion warnings

oontvoo committed:
Merge branch 'llvm:main' into llvm_telemetry

oontvoo committed:
finish up todos

oontvoo commented on lldb/include/lldb/Core/Debugger.h:
```diff
@@ -754,6 +760,7 @@ class Debugger : public std::enable_shared_from_this<Debugger>,
   uint32_t m_interrupt_requested = 0; ///< Tracks interrupt requests
   std::mutex m_interrupt_mutex;
 
+  std::shared_ptr<LldbTelemeter> m_telemeter;
```
The use of  "shared_ptr" here is to ensure the object can be "shared" with different parts of the code - and not so much about the telemter outliving its debugger.

But I suppose we can make it a unique_ptr and just share the raw pointer around, if you prefer?

labath replied:
Yes, please. If the pointer is non-null (I don't remember if you create a dummy object when telemetry is disabled, or if you keep this null), then also make this a reference.

shared_ptr is for sharing the *ownership* of an object. There's nothing wrong with handing out a non-owning reference/pointer to other code (if we can ensure that object is not destroyed in the mean time, which usually means that the other code should not persist that pointer anywhere).

oontvoo replied:
done

oontvoo commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -733,12 +736,20 @@ void Debugger::InstanceInitialize() {
+  TelemetryEventStats init_stats(start_time, std::chrono::steady_clock::now());
+  llvm::telemetry::TelemetryInfo entry;
+  entry.stats = {start_time, std::chrono::steady_clock::now()};
+  debugger_sp->m_telemeter->LogStartup(
```
done

oontvoo commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -952,6 +964,18 @@ void Debugger::Clear() {
+    entry.stats = {/*start_session*/ stats.m_start,
+                   /*end_session*/ std::chrono::steady_clock::now()};
+    m_telemeter->LogExit(
+        HostInfo::GetProgramFileSpec().GetPathAsConstString().GetCString(),
```
done

oontvoo committed:
[lldb]Implement LLDB Telemetry :
Provide the concrete implementation for telemetry in LLDB.

(This is follow-up patch to PR/102323)

oontvoo committed:
change the data type from json to structureddata

oontvoo commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -682,17 +683,51 @@ PacketStatus DAP::GetNextObject(llvm::json::Object &object) {
+    // then we serialize it to a string,
+    // and finally call SBStructuredData::SetFromJSON(string).
+    //
+    // TODO: This seems unnecessarily complex. Ideally, we should be able to
```
@labath Hi Pavel, do you have any suggestion for avoiding this round-about?

labath replied:
Not exactly, since "(SB)StructuredData" is not really JSON, but an abstract representation of structured data (which "happens" to look a lot like json).

The only alternative I see is to add the ability to construct/mutate SBStructuredData directly (all the existing APIs are just getters, probably because our structured data usually flows in the other direction). I think this kind of functionality would make sense from an API perspective, but maybe it's not worth doing it for the sake of this patch.

oontvoo replied:
Thanks - i'll resolve this for now. we can revisit this later

oontvoo committed:
Rework the default impl of Telemetry in LLDB as "Plugin" to allow for testing

oontvoo committed:
Merge branch 'llvm:main' into telemetry_shared

labath reviewed (commented):
I think it'd be good to separate the general mechanics of creating and managing a telemetry (telemeter) instance from the details of logging individual telemetry entries. I.e. one PR for the general infrastructure (maybe include one simple (startup?) telemetry entry, if you think it's necessary to make a point), and then one PR for each of the entries.

The plugin infrastructure seems very.. baroque. Why do we need the intermediate TelemetryVendor class? And why a whole list of them?

The reason we support registering more than one e.g. SymbolFile plugin is that we can choose which one to use at runtime (based on which file the user opens, etc.). Do we want to do something like that here? What would that even mean?

To create a basic indirection (so that the vendor implementation can be decoupled from the rest of lldb), it sounds to me like all one need is one function pointer (the factory function for the "telemeter"), a way to set it from inside the plugin (`PluginManager::SetTelemetryPlugin` ?) and then a way to call it (another function?)

oontvoo commented on lldb/include/lldb/lldb-enumerations.h:
```diff
@@ -257,8 +257,8 @@ enum StopReason {
 /// Command Return Status Types.
-enum ReturnStatus {
-  eReturnStatusInvalid,
+enum ReturnStatus : int {
```
This is needed to serialize the ReturnStatus from a command.

oontvoo committed:
UPdate LLDB implementation to match changes in LLVM's telemetry

oontvoo committed:
fix vtable issue

oontvoo committed:
more renaming for consistency

oontvoo committed:
remove unused

oontvoo committed:
addressed review comments

oontvoo committed:
formatting
